### ISSUE TITLE
Prevent writing to huge partitions

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#12995]
We've seen cases where a partition grew to an extremely large size (single digit TBs), eventually causing issues like running out of disk space when compacting it, adding nodes to the cluster helped alleviate the problem by providing more storage capacity, but then std::bad_alloc prevented cleanup compaction (that needs to be fixed independently in any case).

This issue calls for preventing writes (but not DELETEs) to partitions that cross a certain threshold (could be a function of X times `compaction_large_partition_warning_threshold_mb`, or an optional configurable value).

This is not straight forward as some replicas may already detect the partition as large locally, but some may lag behind if the partition is broken into several sstables, where the part in each of them is below the threshold.

**Comment by avikivity** [comment]
> We've seen cases where a partition grew to an extremely large size (single digit TBs), eventually causing issues like running out of disk space when compacting it, adding nodes to the cluster helped alleviate the problem by providing more storage capacity, but then std::bad_alloc prevented cleanup compaction (that needs to be fixed independently in any case).
> 
> This issue calls for preventing writes (but not DELETEs) to partitions that cross a certain threshold (could be a function of X times `compaction_large_partition_warning_threshold_mb`, or an optional configurable value).
> 
> This is not straight forward as some replicas may already detect the partition as large locally, but some may lag behind if the partition is broken into several sstables, where the part in each of them is below the threshold.

I think it's okay to have a fuzzy limit. It's far from perfect, but it's much better than to allow unbounded growth.

**Comment by avikivity** [comment]
One challenge is to detect those large partitions during startup. We might need to do this by scanning the summary. Another option is to store more information in the Scylla.db compoent.

**Comment by vladzcloudius** [comment]
This issue has a very confusing and IMO incorrect statement. I don't think we ever saw a situation when huge partition in its own caused us running of disk space.

What we did see is that huge partition caused us running out of memory (RAM) . And the reasons for that are numerous.

So, instead of trying to prevent the customer from hitting these issues we would rather fix them instead.

@bhalevy could you, please, at least update the issue description so that it correctly describes the problem you are suggesting to work around?

**Comment by bhalevy** [comment]
> This issue has a very confusing and IMO incorrect statement. I don't think we ever saw a situation when huge partition in its own caused us running of disk space.

It might, since even with ICS we don't break sstable fragments on row boundary yet, and compacting a single fragment out of each run still has a temporary space overhead of 100%, so we need free storage capacity of at least the size of the huge partition in order to compact it.

> 
> What we did see is that huge partition caused us running out of memory (RAM) . And the reasons for that are numerous.
> 
> So, instead of trying to prevent the customer from hitting these issues we would rather fix them instead.
> 

I agree we need to address the failures, but we should also help the customer not to fall into this pothole.
It's less about preventing the errors, but rather preventing the anti-pattern.

> @bhalevy could you, please, at least update the issue description so that it correctly describes the problem you are suggesting to work around?

updated.

**Comment by nyh** [comment]
> This issue calls for preventing writes (but not DELETEs) to partitions that cross a certain threshold (could be a function of X times `compaction_large_partition_warning_threshold_mb`, or an optional configurable value).

We have a similar issue, https://github.com/scylladb/scylladb/issues/10070, asking to limit writes to a **collection** which  grew too large. That issue was about collections but many of the ideas raised there are relevant for the case of partitions too. 

> This is not straight forward as some replicas may already detect the partition as large locally, but some may lag behind if the partition is broken into several sstables, where the part in each of them is below the threshold.

Right. I mentioned this problem in #10070 too. I never had a good solution to it.

**Comment by denesb** [comment]
I think in general, anything that makes it into one of our large* tables, we might consider watching it and after a certain limit, stop accepting writes to it.

**Comment by bhalevy** [comment]
> I think in general, anything that makes it into one of our large* tables, we might consider watching it and after a certain limit, stop accepting writes to it.

The problem we should avoid is that some replicas may recognize the oversized entity (cell/row/partition) as such, and others might not, so some replicas will allow the write and some won't, meaning the write will be partially committed and repair will try to sync it to all replicas, which we should allow, since it's already too late.

So the decision to reject the write should be done based on a cluster decision about large.* entities.
Maybe we can gossip about oversized entities to propagate the information in the cluster and reject the writes in the coordinator.

**Comment by avikivity** [comment]
I think it's okay to reject the write in a fuzzy way. Yes it creates some repair noise, but it's tolerable.

One thing we could do is to fail the write if any replica rejects it due to the partition being too large.

An important thing is to clear the large partition status is a partition tombstone is received.

**Comment by vladzcloudius** [comment]
> I think it's okay to reject the write in a fuzzy way. Yes it creates some repair noise, but it's tolerable.
> 
> One thing we could do is to fail the write if any replica rejects it due to the partition being too large.

@avikivity what you are describing is effectively a "write rollback".
Our eventually consistent data model doesn't allow such things.

Consider following scenarios:

Scenario 1: Write key K, RF=3, CL=2
1) Replica A and C accepted the write
2) replica B rejected it. 
3) Coordinator failed the write (due to B rejection - as you suggested)
4.0) key K is read from A and B with CL=2 and read-repaired to B. 
(if we decide to refuse a read-repair)
4.1) key K is read from A and C with CL=2 - data returned.
4.2) key K is read from A and B with CL=2 - nothing is returned (since we would identify a large item and fail the read).

Bottom line - I don't see how we can implement this feature in a useful way without unacceptable performance penalty (suggested by Benny here: https://github.com/scylladb/scylladb/issues/12995#issuecomment-1457795373).


> We have a similar issue, https://github.com/scylladb/scylladb/issues/10070, asking to limit writes to a collection which grew too large.

@nyh Do we have a summary of all relevant ideas from the issue above?

**Comment by mykaul** [comment]
@psarna - I'm not sure if that's part of your TODO list for guardrails, just FYI in case it isn't.

**Comment by nyh** [comment]
> @psarna - I'm not sure if that's part of your TODO list for guardrails, just FYI in case it isn't.

Wrong Piotr, that should be @ptrsmrn :-)

**Comment by mykaul** [comment]
> > @psarna - I'm not sure if that's part of your TODO list for guardrails, just FYI in case it isn't.
> 
> Wrong Piotr, that should be @ptrsmrn :-)

ARGH! Edited.

**Comment by ptrsmrn** [comment]
@mykaul We don't have anything regarding partition size in the [guardrails spec](https://docs.google.com/document/d/11OI8SktNVZKfwo_KQB6mmHUMfXYeevGm4whSYxmgE_A/edit?usp=sharing) - I will add this idea - but we have a "Number of items per Collection" guardrail proposed (see my comment here: https://github.com/scylladb/scylladb/issues/10070#issuecomment-1534772807)

**Comment by avikivity** [comment]
> > I think it's okay to reject the write in a fuzzy way. Yes it creates some repair noise, but it's tolerable.
> > One thing we could do is to fail the write if any replica rejects it due to the partition being too large.
> 
> @avikivity what you are describing is effectively a "write rollback". Our eventually consistent data model doesn't allow such things.
> 
> Consider following scenarios:
> 
> Scenario 1: Write key K, RF=3, CL=2
> 
> 1. Replica A and C accepted the write
> 2. replica B rejected it.
> 3. Coordinator failed the write (due to B rejection - as you suggested)

The coordinator would respond (with ACK) after 1. It's only possible to reject before CL is reached.


>    4.0) key K is read from A and B with CL=2 and read-repaired to B.
>    (if we decide to refuse a read-repair)
>    4.1) key K is read from A and C with CL=2 - data returned.
>    4.2) key K is read from A and B with CL=2 - nothing is returned (since we would identify a large item and fail the read).
> 

We'd only fail writes, not reads.


> Bottom line - I don't see how we can implement this feature in a useful way without unacceptable performance penalty (suggested by Benny here: [#12995 (comment)](https://github.com/scylladb/scylladb/issues/12995#issuecomment-1457795373)).
> 
> > We have a similar issue, #10070, asking to limit writes to a collection which grew too large.

**Comment by vladzcloudius** [comment]
> > > I think it's okay to reject the write in a fuzzy way. Yes it creates some repair noise, but it's tolerable.
> > > One thing we could do is to fail the write if any replica rejects it due to the partition being too large.
> > 
> > 
> > @avikivity what you are describing is effectively a "write rollback". Our eventually consistent data model doesn't allow such things.
> > Consider following scenarios:
> > Scenario 1: Write key K, RF=3, CL=2
> > 
> > 1. Replica A and C accepted the write
> > 2. replica B rejected it.
> > 3. Coordinator failed the write (due to B rejection - as you suggested)
> 
> The coordinator would respond (with ACK) after 1. It's only possible to reject before CL is reached.

If A and C respond first, yes, but you can assume A and B responded first.
If Coordinator responds with ACK - it's much-much worse because we will never be able to achieve a consistency on a write that we have acknowledged. - which is another reason why this idea is not great.

> 
> > 4.0) key K is read from A and B with CL=2 and read-repaired to B.
> > (if we decide to refuse a read-repair)
> > 4.1) key K is read from A and C with CL=2 - data returned.
> > 4.2) key K is read from A and B with CL=2 - nothing is returned (since we would identify a large item and fail the read).
> 
> We'd only fail writes, not reads.

And what would you do in the situation from my example? Allow read-repair? 
If "no", what data you would return in 4.2? 
If "yes" - this kills the purpose of the "protection" in question.


> 
> > Bottom line - I don't see how we can implement this feature in a useful way without unacceptable performance penalty (suggested by Benny here: [#12995 (comment)](https://github.com/scylladb/scylladb/issues/12995#issuecomment-1457795373)).
> > > We have a similar issue, #10070, asking to limit writes to a collection which grew too large.

**Comment by avikivity** [comment]
> > > > I think it's okay to reject the write in a fuzzy way. Yes it creates some repair noise, but it's tolerable.
> > > > One thing we could do is to fail the write if any replica rejects it due to the partition being too large.
> > > 
> > > 
> > > @avikivity what you are describing is effectively a "write rollback". Our eventually consistent data model doesn't allow such things.
> > > Consider following scenarios:
> > > Scenario 1: Write key K, RF=3, CL=2
> > > 
> > > 1. Replica A and C accepted the write
> > > 2. replica B rejected it.
> > > 3. Coordinator failed the write (due to B rejection - as you suggested)
> > 
> > 
> > The coordinator would respond (with ACK) after 1. It's only possible to reject before CL is reached.
> 
> If A and C respond first, yes, 

Your example (step 1) had A and C responding first. 

> but you can assume A and B responded first for my example. If A and C respond first and Coordinator responds with ACK - it's much-much worse because we will never be able to achieve a consistency on a write that we have acknowledged. - which is another reason why this idea is not great.

We'll achieve consistency once the user trims down the partition.


> 
> > > 4.0) key K is read from A and B with CL=2 and read-repaired to B.
> > > (if we decide to refuse a read-repair)
> > > 4.1) key K is read from A and C with CL=2 - data returned.
> > > 4.2) key K is read from A and B with CL=2 - nothing is returned (since we would identify a large item and fail the read).
> > 
> > 
> > We'd only fail writes, not reads.
> 
> And what would you do in the situation from my example? Allow read-repair? If "no", what data you would return in 4.2? If "yes" - this kills the purpose of the "protection" in question since read-repair is going to generate a large partition on B.


That's a good point. I guess we should fail read-repair (and therefore the read too).

**Comment by vladzcloudius** [comment]
> > > > > I think it's okay to reject the write in a fuzzy way. Yes it creates some repair noise, but it's tolerable.
> > > > > One thing we could do is to fail the write if any replica rejects it due to the partition being too large.
> > > > 
> > > > 
> > > > @avikivity what you are describing is effectively a "write rollback". Our eventually consistent data model doesn't allow such things.
> > > > Consider following scenarios:
> > > > Scenario 1: Write key K, RF=3, CL=2
> > > > 
> > > > 1. Replica A and C accepted the write
> > > > 2. replica B rejected it.
> > > > 3. Coordinator failed the write (due to B rejection - as you suggested)
> > > 
> > > 
> > > The coordinator would respond (with ACK) after 1. It's only possible to reject before CL is reached.
> > 
> > 
> > If A and C respond first, yes,
> 
> Your example (step 1) had A and C responding first.

I've already clarified that the intent was that A (or C) and B respond first in my previous post. In the initial example description I didn't mean to specify the order of replica responses order - just what they were. 

> 
> > but you can assume A and B responded first for my example. If A and C respond first and Coordinator responds with ACK - it's much-much worse because we will never be able to achieve a consistency on a write that we have acknowledged. - which is another reason why this idea is not great.
> 
> We'll achieve consistency once the user trims down the partition.

This is going to open a worm can of all kinds of corner cases similar to my example above.

> 
> > > > 4.0) key K is read from A and B with CL=2 and read-repaired to B.
> > > > (if we decide to refuse a read-repair)
> > > > 4.1) key K is read from A and C with CL=2 - data returned.
> > > > 4.2) key K is read from A and B with CL=2 - nothing is returned (since we would identify a large item and fail the read).
> > > 
> > > 
> > > We'd only fail writes, not reads.
> > 
> > 
> > And what would you do in the situation from my example? Allow read-repair? If "no", what data you would return in 4.2? If "yes" - this kills the purpose of the "protection" in question since read-repair is going to generate a large partition on B.
> 
> That's a good point. I guess we should fail read-repair (and therefore the read too).

Exactly. Which brings us to the read monotonicity issue I demonstrated in my example.

**Comment by bhalevy** [comment]
@nyh wrote:
> I also don't know how to estimate the total size of a partition that lives in multiple sstables. Remember that if you see 0.5 GB of a partition in one sstable and 0.5 GB in the same partition in a second sstable, it might be different rows so the total will be 1GB, or it might be an overwrite of the same rows (or tombstones) and the total will be 0.5 GB. I don't think we have an efficient way to tell these cases apart which is not compaction.

Detecting the large partition outside of compaction may be possible on the read path if we scan all of it, but
I really don't think it's worth it.  Detecting as we do today and preventing new writes into it will still provide important benefits.

The main question is where to prevent those writes.
As I stated earlier on this issue, rejecting the writes on replicas is problematic when some replicas detected the partition as large, but not all of them, and you end up with a fuzzy write failure that you still need to repair.
My proposal (https://github.com/scylladb/scylladb/issues/12995#issuecomment-1457795373), is to use gossip to broadcast the state of large partitions and reject writes to them on the coordinators, based on the gossip state.

This sill might have issues with reaction delay and maybe false positive if we don't update this state soon enough
after data is deleted, but we can have ways around it like maybe TTLing the state and refreshing it periodically,
and/or adding an api to clear the state or override the guardrails for a particular partition that is critical to the application, despite being too large.

**Comment by avikivity** [comment]
APIs to clear large partition entries sound like the cure is worse than the disease.

**Comment by asias** [comment]
> @nyh wrote:
> 
> > I also don't know how to estimate the total size of a partition that lives in multiple sstables. Remember that if you see 0.5 GB of a partition in one sstable and 0.5 GB in the same partition in a second sstable, it might be different rows so the total will be 1GB, or it might be an overwrite of the same rows (or tombstones) and the total will be 0.5 GB. I don't think we have an efficient way to tell these cases apart which is not compaction.
> 
> Detecting the large partition outside of compaction may be possible on the read path if we scan all of it, but I really don't think it's worth it. Detecting as we do today and preventing new writes into it will still provide important benefits.
> 
> The main question is where to prevent those writes. As I stated earlier on this issue, rejecting the writes on replicas is problematic when some replicas detected the partition as large, but not all of them, and you end up with a fuzzy write failure that you still need to repair. My proposal ([#12995 (comment)](https://github.com/scylladb/scylladb/issues/12995#issuecomment-1457795373)), is to use gossip to broadcast the state of large partitions and reject writes to them on the coordinators, based on the gossip state.
> 
> This sill might have issues with reaction delay and maybe false positive if we don't update this state soon enough after data is deleted, but we can have ways around it like maybe TTLing the state and refreshing it periodically, and/or adding an api to clear the state or override the guardrails for a particular partition that is critical to the application, despite being too large.

It depends on how many such large partitions we are going to have.  if it is too many,  Gossip is not a good fit.  

But I have an idea to solve the out of date big partition info issue. 

1) Get large partition list from the broadcast, e.g., from gossip or any other mechanisms

2) if it is a large partition from the broadcast info, send a dedicated verb to any or all of replica sets, to get a confirmation if the partition is indeed still large.  If any of the query returns true, we reject the large partition write. 

The extra verb on stage 2 adds latency but it has a high chance the partition is indeed a big partition, so it is not that bad.

This guarantees that we do not reject a write incorrectly after a user has deleted the partition.

**Comment by bhalevy** [comment]
Instead of gossip we can have the replica return information on the write response about different violations of limits.
This could be a violation of soft or hard limit for single partitions or violations of the free disk space limit for #14067.

After learning about this from the replicas, the coordinator can establish a state on its side to reject the *next* write based on the previous responses.  To prevent inconsistencies, I suggest that the writes will succeed on the replicas, but the warning will be returned with the result.

IIUC, we can propagate the warnings to the client as well over CQL once we cross the soft limit on any of the replicas so the user can take action before things worsen to the point that the hard limit is crossed.

As for validating the guardrails state.  I think @asias idea of querying the replicas about their state once we have a guardrails event sounds feasible.  In case of large partitions, the coordinator can read the data from the replica using CQL by querying the system.large_partitions table there.  For disk-space statistics we can add gossip application states for total and used space. I think it's a great fit for gossip.

**Comment by bhalevy** [comment]
@ptrsmrn since this feature has a lot in common to out-of-space prevention I suggest reassigning it to @paszkow.
Please confirm.

**Comment by ptrsmrn** [comment]
> [@ptrsmrn](https://github.com/ptrsmrn) since this feature has a lot in common to out-of-space prevention I suggest reassigning it to [@paszkow](https://github.com/paszkow). Please confirm.

@bhalevy sure, reassigned to @paszkow

**Comment by bhalevy** [comment]
@paszkow please add to 2026.1 planned contents
Let's make sure it is indeed realistic

**Comment by bhalevy** [comment]
I'd like to have a plan for implementing those guardrails incrementally and have and epic to track those deliverables.
I am not sure it can all be done within a single release.
For example:
* Reject writes of large cells in the coordinator
  * This for atomic cells (could be long blobs or strings) and frozen collections
* Reject writes to large data entities in the replica
  * Large collections
  * Large rows
  * Partitions with too many rows
    * Need to decide what to do if the reason for many rows is tombstone accumulation.  On one hand we want to allow deletions to reduce the partition size, but OTOH, we should ensure tombstones can be garbage collected and purged/
  * Large partitions

**Comment by ptrsmrn** [comment]
@mykaul how does that relate to https://scylladb.atlassian.net/browse/SCYLLADB-254

**Comment by mykaul** [comment]
> [@mykaul](https://github.com/mykaul) how does that relate to https://scylladb.atlassian.net/browse/SCYLLADB-254

@ptrsmrn - not sure I understand this question. This prevention, as imperfect as it may be, is a critical guardrail. MUCH more important than configuration guardrail (although it's related - you want to set a configurable limit on what 'large' cell/partition/whatnot' is. FAR more important than using or not using SimpleStrategy.

**Comment by ptrsmrn** [comment]
@mykaul who should work on completing this gh issue? Does resuming guardrails work in [scylladb.atlassian.net/browse/SCYLLADB-254](https://scylladb.atlassian.net/browse/SCYLLADB-254) relates to this specific gh issue, does it change anything, has any impact, from organizational perspective? Core-frontend has a separate effort for guardrails, and so do @paszkow 's team. Should they continue as planned despite us resuming guardrails work? I'm asking if our jira is anyhow related to this gh issue.
